diff --git a/src/pke/lib/scheme/ckksrns/ckksrns-advancedshe.cpp b/src/pke/lib/scheme/ckksrns/ckksrns-advancedshe.cpp
index 068930f7..e2a8d368 100644
--- a/src/pke/lib/scheme/ckksrns/ckksrns-advancedshe.cpp
+++ b/src/pke/lib/scheme/ckksrns/ckksrns-advancedshe.cpp
@@ -36,6 +36,8 @@ CKKS implementation. See https://eprint.iacr.org/2020/1118 for details.
 #include "cryptocontext.h"
 #include "scheme/ckksrns/ckksrns-cryptoparameters.h"
 #include "scheme/ckksrns/ckksrns-advancedshe.h"
+
+#include "scheme/gen-cryptocontext-params-defaults.h"
 #include "scheme/ckksrns/ckksrns-utils.h"
 #include "schemebase/base-scheme.h"
 
@@ -87,61 +89,95 @@ Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalMultMany(const std::vector<Cipherte
 
 template <typename VectorDataType>
 Ciphertext<DCRTPoly> internalEvalLinearWSum(const std::vector<ReadOnlyCiphertext<DCRTPoly>>& ciphertexts,
-                                            const std::vector<VectorDataType>& constants) {
+                                            const std::vector<VectorDataType>& constants, int32_t target_level) {
     const uint32_t limit = ciphertexts.size();
     std::vector<Ciphertext<DCRTPoly>> cts(limit);
     for (uint32_t i = 0; i < limit; ++i)
         cts[i] = ciphertexts[i]->Clone();
-    return internalEvalLinearWSumMutable(cts, constants);
+    return internalEvalLinearWSumMutable(cts, constants, target_level);
 }
 
 template <typename VectorDataType>
 Ciphertext<DCRTPoly> internalEvalLinearWSumMutable(std::vector<Ciphertext<DCRTPoly>>& ciphertexts,
-                                                   const std::vector<VectorDataType>& constants) {
+                                                   const std::vector<VectorDataType>& constants,
+                                                   int32_t target_level = -1) {
     const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(ciphertexts[0]->GetCryptoParameters());
 
     auto cc = ciphertexts[0]->GetCryptoContext();
 
     const uint32_t limit = ciphertexts.size();
 
-    if (cryptoParams->GetScalingTechnique() != FIXEDMANUAL) {
-        // Check to see if input ciphertexts are of same level
-        // and adjust if needed to the max level among them
-        uint32_t maxLevel = ciphertexts[0]->GetLevel();
-        uint32_t maxIdx   = 0;
+    if (target_level == -1) {
+        if (cryptoParams->GetScalingTechnique() != FIXEDMANUAL) {
+            // Check to see if input ciphertexts are of same level
+            // and adjust if needed to the max level among them
+            uint32_t maxLevel = ciphertexts[0]->GetLevel();
+            uint32_t maxIdx   = 0;
+            for (uint32_t i = 1; i < limit; ++i) {
+                if ((ciphertexts[i]->GetLevel() > maxLevel) ||
+                    ((ciphertexts[i]->GetLevel() == maxLevel) && (ciphertexts[i]->GetNoiseScaleDeg() == 2))) {
+                    maxLevel = ciphertexts[i]->GetLevel();
+                    maxIdx   = i;
+                    }
+            }
+
+            auto algo = cc->GetScheme();
+            for (uint32_t i = 0; i < maxIdx; ++i)
+                algo->AdjustLevelsAndDepthInPlace(ciphertexts[i], ciphertexts[maxIdx]);
+            for (uint32_t i = maxIdx + 1; i < limit; ++i)
+                algo->AdjustLevelsAndDepthInPlace(ciphertexts[i], ciphertexts[maxIdx]);
+
+            uint32_t compositeDegree = cryptoParams->GetCompositeDegree();
+            if (ciphertexts[maxIdx]->GetNoiseScaleDeg() == 2) {
+                for (uint32_t i = 0; i < limit; ++i)
+                    algo->ModReduceInternalInPlace(ciphertexts[i], compositeDegree);
+            }
+        }
+
+        cc->EvalMultInPlace(ciphertexts[0], constants[0]);
+        for (uint32_t i = 1; i < limit; ++i) {
+            cc->EvalMultInPlace(ciphertexts[i], constants[i]);
+            cc->EvalAddInPlaceNoCheck(ciphertexts[0], ciphertexts[i]);
+        }
+        cc->ModReduceInPlace(ciphertexts[0]);
+        return ciphertexts[0];
+    } else {
+        //std::cout << "new internalEvalLinearWSumMutable" << std::endl;
+        size_t maxLevel = ciphertexts[0]->GetLevel();
+        if (ciphertexts[0]->GetNoiseScaleDeg() == 2)
+            OPENFHE_THROW("Not modulo reduced");
+        //uint32_t maxIdx   = 0;
         for (uint32_t i = 1; i < limit; ++i) {
             if ((ciphertexts[i]->GetLevel() > maxLevel) ||
                 ((ciphertexts[i]->GetLevel() == maxLevel) && (ciphertexts[i]->GetNoiseScaleDeg() == 2))) {
                 maxLevel = ciphertexts[i]->GetLevel();
-                maxIdx   = i;
-            }
+                //maxIdx   = i;
+                }
+
+            if (ciphertexts[i]->GetNoiseScaleDeg() == 2)
+                OPENFHE_THROW("Not modulo reduced");
         }
 
-        auto algo = cc->GetScheme();
-        for (uint32_t i = 0; i < maxIdx; ++i)
-            algo->AdjustLevelsAndDepthInPlace(ciphertexts[i], ciphertexts[maxIdx]);
-        for (uint32_t i = maxIdx + 1; i < limit; ++i)
-            algo->AdjustLevelsAndDepthInPlace(ciphertexts[i], ciphertexts[maxIdx]);
+        if((int32_t)maxLevel > target_level) {
+            OPENFHE_THROW("Target level is too low (too many towers requested)");
+        }
 
-        uint32_t compositeDegree = cryptoParams->GetCompositeDegree();
-        if (ciphertexts[maxIdx]->GetNoiseScaleDeg() == 2) {
-            for (uint32_t i = 0; i < limit; ++i)
-                algo->ModReduceInternalInPlace(ciphertexts[i], compositeDegree);
+        cc->EvalMultInPlace(ciphertexts[0], constants[0], target_level);
+        auto out = ciphertexts[0]->Clone();
+        for (uint32_t i = 1; i < limit; ++i) {
+            cc->EvalMultInPlace(ciphertexts[i], constants[i], target_level);
+            cc->EvalAddInPlace(out, ciphertexts[i]);
         }
+        cc->ModReduceInPlace(out);
+        return out;
     }
 
-    cc->EvalMultInPlace(ciphertexts[0], constants[0]);
-    for (uint32_t i = 1; i < limit; ++i) {
-        cc->EvalMultInPlace(ciphertexts[i], constants[i]);
-        cc->EvalAddInPlaceNoCheck(ciphertexts[0], ciphertexts[i]);
-    }
-    cc->ModReduceInPlace(ciphertexts[0]);
-    return ciphertexts[0];
 }
 
 template <typename VectorDataType>
 Ciphertext<DCRTPoly> EvalPartialLinearWSum(const std::vector<Ciphertext<DCRTPoly>>& ciphertexts,
-                                           const std::vector<VectorDataType>& constants, uint32_t limit = 0) {
+                                           const std::vector<VectorDataType>& constants, uint32_t limit = 0,
+                                           int32_t target_level = -1, bool rescale = true) {
     if (0 == limit)
         limit = ciphertexts.size();
 
@@ -149,73 +185,146 @@ Ciphertext<DCRTPoly> EvalPartialLinearWSum(const std::vector<Ciphertext<DCRTPoly
 
     auto cc = ciphertexts[0]->GetCryptoContext();
 
-    std::vector<Ciphertext<DCRTPoly>> cts(limit);
-    if (cryptoParams->GetScalingTechnique() != FIXEDMANUAL) {
-        cts[0] = ciphertexts[0]->Clone();
-        // Check to see if input ciphertexts are of same level
-        // and adjust if needed to the max level among them
-        uint32_t maxLevel = cts[0]->GetLevel();
-        uint32_t maxIdx   = 0;
-        for (uint32_t i = 1; i < limit; ++i) {
-            cts[i] = ciphertexts[i]->Clone();
-            if ((cts[i]->GetLevel() > maxLevel) ||
-                ((cts[i]->GetLevel() == maxLevel) && (cts[i]->GetNoiseScaleDeg() == 2))) {
-                maxLevel = cts[i]->GetLevel();
-                maxIdx   = i;
+    if (target_level == -1) {
+        std::vector<Ciphertext<DCRTPoly>> cts(limit);
+        if (cryptoParams->GetScalingTechnique() != FIXEDMANUAL) {
+            cts[0] = ciphertexts[0]->Clone();
+            // Check to see if input ciphertexts are of same level
+            // and adjust if needed to the max level among them
+            uint32_t maxLevel = cts[0]->GetLevel();
+            uint32_t maxIdx   = 0;
+            for (uint32_t i = 1; i < limit; ++i) {
+                cts[i] = ciphertexts[i]->Clone();
+                if ((cts[i]->GetLevel() > maxLevel) ||
+                    ((cts[i]->GetLevel() == maxLevel) && (cts[i]->GetNoiseScaleDeg() == 2))) {
+                    maxLevel = cts[i]->GetLevel();
+                    maxIdx   = i;
+                    }
+            }
+
+            auto algo = cc->GetScheme();
+            auto& ctm = cts[maxIdx];
+            for (uint32_t i = 0; i < maxIdx; ++i)
+                algo->AdjustLevelsAndDepthInPlace(cts[i], ctm);
+            for (uint32_t i = maxIdx + 1; i < limit; ++i)
+                algo->AdjustLevelsAndDepthInPlace(cts[i], ctm);
+
+            uint32_t compositeDegree = cryptoParams->GetCompositeDegree();
+            if (ctm->GetNoiseScaleDeg() == 2) {
+                for (uint32_t i = 0; i < limit; ++i)
+                    algo->ModReduceInternalInPlace(cts[i], compositeDegree);
             }
         }
+        else {
+            for (uint32_t i = 0; i < limit; ++i)
+                cts[i] = ciphertexts[i]->Clone();
+        }
 
-        auto algo = cc->GetScheme();
-        auto& ctm = cts[maxIdx];
-        for (uint32_t i = 0; i < maxIdx; ++i)
-            algo->AdjustLevelsAndDepthInPlace(cts[i], ctm);
-        for (uint32_t i = maxIdx + 1; i < limit; ++i)
-            algo->AdjustLevelsAndDepthInPlace(cts[i], ctm);
 
-        uint32_t compositeDegree = cryptoParams->GetCompositeDegree();
-        if (ctm->GetNoiseScaleDeg() == 2) {
-            for (uint32_t i = 0; i < limit; ++i)
-                algo->ModReduceInternalInPlace(cts[i], compositeDegree);
+        cc->EvalMultInPlace(cts[0], constants[1]);
+        for (uint32_t i = 1; i < limit; ++i) {
+            cc->EvalMultInPlace(cts[i], constants[i + 1]);
+            cc->EvalAddInPlaceNoCheck(cts[0], cts[i]);
         }
+        cc->ModReduceInPlace(cts[0]);
+        return cts[0];
+    } else {
+        //std::cout << "new internalEvalLinearWSum" << std::endl;
+        size_t maxLevel = ciphertexts[0]->GetLevel();
+        if (ciphertexts[0]->GetNoiseScaleDeg() == 2)
+            OPENFHE_THROW("Not modulo reduced");
+        //uint32_t maxIdx   = 0;
+        for (uint32_t i = 1; i < limit; ++i) {
+            if ((ciphertexts[i]->GetLevel() > maxLevel) ||
+                ((ciphertexts[i]->GetLevel() == maxLevel) && (ciphertexts[i]->GetNoiseScaleDeg() == 2))) {
+                maxLevel = ciphertexts[i]->GetLevel();
+                //maxIdx   = i;
+                }
+
+            if (ciphertexts[i]->GetNoiseScaleDeg() == 2)
+                OPENFHE_THROW("Not modulo reduced");
+        }
+
+        auto out = cc->EvalMult(ciphertexts[0], constants[1], target_level);
+        for (uint32_t i = 1; i < limit; ++i) {
+            cc->EvalAddInPlace(out, cc->EvalMult(ciphertexts[i], constants[i+1], target_level));
+        }
+        if (rescale) cc->ModReduceInPlace(out);
+        return out;
     }
-    else {
-        for (uint32_t i = 0; i < limit; ++i)
-            cts[i] = ciphertexts[i]->Clone();
-    }
+}
+
+template <typename VectorDataType>
+Ciphertext<DCRTPoly> EvalPartialLinearWSumWithBias(const std::vector<Ciphertext<DCRTPoly>>& ciphertexts,
+                                                   const std::vector<VectorDataType>& constants, VectorDataType bias,
+                                                   uint32_t limit = 0, int32_t target_level = -1, bool rescale = true) {
+    if (0 == limit)
+        limit = ciphertexts.size();
+
+    const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(ciphertexts[0]->GetCryptoParameters());
+
+    auto cc = ciphertexts[0]->GetCryptoContext();
 
-    cc->EvalMultInPlace(cts[0], constants[1]);
-    for (uint32_t i = 1; i < limit; ++i) {
-        cc->EvalMultInPlace(cts[i], constants[i + 1]);
-        cc->EvalAddInPlaceNoCheck(cts[0], cts[i]);
+    if (target_level == -1) {
+        OPENFHE_THROW("Not implemented for BASELINE cheby or other uses");
+    } else {
+        //std::cout << "new internalEvalLinearWSum" << std::endl;
+        size_t maxLevel = ciphertexts[0]->GetLevel();
+        if (ciphertexts[0]->GetNoiseScaleDeg() == 2)
+            OPENFHE_THROW("Not modulo reduced");
+        //uint32_t maxIdx   = 0;
+        for (uint32_t i = 1; i < limit; ++i) {
+            if ((ciphertexts[i]->GetLevel() > maxLevel) ||
+                ((ciphertexts[i]->GetLevel() == maxLevel) && (ciphertexts[i]->GetNoiseScaleDeg() == 2))) {
+                maxLevel = ciphertexts[i]->GetLevel();
+                //maxIdx   = i;
+                }
+
+            if (ciphertexts[i]->GetNoiseScaleDeg() == 2)
+                OPENFHE_THROW("Not modulo reduced");
+        }
+
+        auto out = cc->EvalMult(ciphertexts[0], constants[1], target_level);
+        for (uint32_t i = 1; i < limit; ++i) {
+            if (IsNotEqualZero(constants[i+1]))
+                cc->EvalAddInPlace(out, cc->EvalMult(ciphertexts[i], constants[i+1], target_level));
+        }
+        cc->EvalAddInPlace(out, bias);
+        if (rescale) cc->ModReduceInPlace(out);
+        return out;
     }
-    cc->ModReduceInPlace(cts[0]);
-    return cts[0];
 }
 
 Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSum(std::vector<ReadOnlyCiphertext<DCRTPoly>>& ciphertexts,
-                                                        const std::vector<int64_t>& constants) const {
-    return internalEvalLinearWSum(ciphertexts, constants);
+                                                        const std::vector<int64_t>& constants,
+                                                        int32_t target_level) const {
+    return internalEvalLinearWSum(ciphertexts, constants, target_level);
 }
 Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSum(std::vector<ReadOnlyCiphertext<DCRTPoly>>& ciphertexts,
-                                                        const std::vector<double>& constants) const {
-    return internalEvalLinearWSum(ciphertexts, constants);
+                                                        const std::vector<double>& constants,
+                                                        int32_t target_level) const {
+    return internalEvalLinearWSum(ciphertexts, constants, target_level);
 }
 Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSum(std::vector<ReadOnlyCiphertext<DCRTPoly>>& ciphertexts,
-                                                        const std::vector<std::complex<double>>& constants) const {
-    return internalEvalLinearWSum(ciphertexts, constants);
+                                                        const std::vector<std::complex<double>>& constants,
+                                                        int32_t target_level) const {
+    return internalEvalLinearWSum(ciphertexts, constants, target_level);
 }
 
 Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSumMutable(std::vector<Ciphertext<DCRTPoly>>& ciphertexts,
-                                                               const std::vector<int64_t>& constants) const {
-    return internalEvalLinearWSumMutable(ciphertexts, constants);
+                                                               const std::vector<int64_t>& constants,
+                                                               int32_t target_level) const {
+    return internalEvalLinearWSumMutable(ciphertexts, constants, target_level);
 }
 Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSumMutable(std::vector<Ciphertext<DCRTPoly>>& ciphertexts,
-                                                               const std::vector<double>& constants) const {
-    return internalEvalLinearWSumMutable(ciphertexts, constants);
+                                                               const std::vector<double>& constants,
+                                                               int32_t target_level) const {
+    return internalEvalLinearWSumMutable(ciphertexts, constants, target_level);
 }
-Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSumMutable(
-    std::vector<Ciphertext<DCRTPoly>>& ciphertexts, const std::vector<std::complex<double>>& constants) const {
-    return internalEvalLinearWSumMutable(ciphertexts, constants);
+Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalLinearWSumMutable(std::vector<Ciphertext<DCRTPoly>>& ciphertexts,
+                                                               const std::vector<std::complex<double>>& constants,
+                                                               int32_t target_level) const {
+    return internalEvalLinearWSumMutable(ciphertexts, constants, target_level);
 }
 
 //------------------------------------------------------------------------------
@@ -567,6 +676,8 @@ Ciphertext<DCRTPoly> AdvancedSHECKKSRNS::EvalPolyPS(ConstCiphertext<DCRTPoly>& x
 //------------------------------------------------------------------------------
 // EVAL CHEBYSHEV SERIES
 //------------------------------------------------------------------------------
+constexpr bool BASELINE = true;
+constexpr bool CHEBY_PRINT = false;
 
 template <typename VectorDataType>
 std::shared_ptr<seriesPowers<DCRTPoly>> internalEvalChebyPolysLinear(ConstCiphertext<DCRTPoly>& x,
@@ -612,11 +723,21 @@ std::shared_ptr<seriesPowers<DCRTPoly>> internalEvalChebyPolysLinear(ConstCipher
         }
     }
 
-    uint32_t compositeDegree =
+    if (!BASELINE) {
+        uint32_t compositeDegree =
         std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
-    for (uint32_t i = 1; i < k; ++i)
-        cc->LevelReduceInPlace(T[i - 1], nullptr, (T[k - 1]->GetLevel() - T[i - 1]->GetLevel()) / compositeDegree);
 
+        for (uint32_t i = 1; i <= k; ++i) {
+            if (T[i-1]->GetNoiseScaleDeg() == 2)
+                cc->GetScheme()->ModReduceInternalInPlace(T[i-1], compositeDegree);
+            //   cc->LevelReduceInPlace(T[i - 1], nullptr, (T[k - 1]->GetLevel() - T[i - 1]->GetLevel()) / compositeDegree);
+        }
+    } else {
+        uint32_t compositeDegree =
+        std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
+        for (uint32_t i = 1; i < k; ++i)
+            cc->LevelReduceInPlace(T[i - 1], nullptr, (T[k - 1]->GetLevel() - T[i - 1]->GetLevel()) / compositeDegree);
+    }
     return std::make_shared<seriesPowers<DCRTPoly>>(std::move(T));
 }
 
@@ -649,7 +770,7 @@ Ciphertext<DCRTPoly> internalEvalChebyshevSeriesLinearWithPrecomp(std::vector<Ci
 template <typename VectorDataType>
 Ciphertext<DCRTPoly> InnerEvalChebyshevPS(ConstCiphertext<DCRTPoly>& x, const std::vector<VectorDataType>& coefficients,
                                           uint32_t k, uint32_t m, const std::vector<Ciphertext<DCRTPoly>>& T,
-                                          const std::vector<Ciphertext<DCRTPoly>>& T2) {
+                                          const std::vector<Ciphertext<DCRTPoly>>& T2, int32_t level_offset = 0) {
     // Compute k*2^{m-1}-k because we use it a lot
     uint32_t k2m2k = k * (1 << (m - 1)) - k;
 
@@ -678,27 +799,36 @@ Ciphertext<DCRTPoly> InnerEvalChebyshevPS(ConstCiphertext<DCRTPoly>& x, const st
         // Evaluate q and s2 at u.
         // If their degrees are larger than k, then recursively apply the Paterson-Stockmeyer algorithm.
         if (Degree(divqr->q) > k) {
-            qu = InnerEvalChebyshevPS(x, divqr->q, k, m - 1, T, T2);
+            qu = InnerEvalChebyshevPS(x, divqr->q, k, m - 1, T, T2, level_offset);
+            if (!BASELINE) cc->ModReduceInPlace(qu);
         }
         else {
-            // dq = k from construction
-            // perform scalar multiplication for all other terms and sum them up if there are non-zero coefficients
-
-            // the highest order coefficient will always be a power of two up to 2^{m-1} because q is "monic" but the Chebyshev rule adds a factor of 2
-            // we don't need to increase the depth by multiplying the highest order coefficient, but instead checking and summing, since we work with m <= 4.
-            qu                   = T[k - 1]->Clone();
-            const uint32_t limit = std::log2(ToReal(divqr->q.back()));
-            for (uint32_t i = 0; i < limit; ++i)
-                cc->EvalAddInPlaceNoCheck(qu, qu);
-
-            // adds the free term (at x^0)
-            cc->EvalAddInPlace(qu, divqr->q.front() / 2.0);
-            // The number of levels of qu is the same as the number of levels of T[k-1] + 1.
-            // Will only get here when m = 2, so the number of levels of qu and T2[m-1] will be the same.
-
-            divqr->q.resize(k);
-            if (uint32_t n = Degree(divqr->q); n > 0)
-                cc->EvalAddInPlace(qu, EvalPartialLinearWSum(T, divqr->q, n));
+            if (BASELINE) {
+                // dq = k from construction
+                // perform scalar multiplication for all other terms and sum them up if there are non-zero coefficients
+
+                // the highest order coefficient will always be a power of two up to 2^{m-1} because q is "monic" but the Chebyshev rule adds a factor of 2
+                // we don't need to increase the depth by multiplying the highest order coefficient, but instead checking and summing, since we work with m <= 4.
+                qu                   = T[k - 1]->Clone();
+                const uint32_t limit = std::log2(ToReal(divqr->q.back()));
+                for (uint32_t i = 0; i < limit; ++i)
+                    cc->EvalAddInPlaceNoCheck(qu, qu); // Change this to integer mult????
+
+                // adds the free term (at x^0)
+                cc->EvalAddInPlace(qu, divqr->q.front() / 2.0);
+                //if (!BASELINE)
+                //    cc->EvalMultInPlace(qu, 1.0, (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset);
+                // The number of levels of qu is the same as the number of levels of T[k-1] + 1.
+                // Will only get here when m = 2, so the number of levels of qu and T2[m-1] will be the same.
+
+                divqr->q.resize(k);
+                if (uint32_t n = Degree(divqr->q); n > 0)
+                    cc->EvalAddInPlace(qu, EvalPartialLinearWSum(T, divqr->q, n, BASELINE ? -1 : (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset));
+            } else {
+                qu = EvalPartialLinearWSumWithBias(T, divqr->q, divqr->q.front()/(VectorDataType)2.0,
+                    divqr->q.size()-1, (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset);
+            }
+            if constexpr (CHEBY_PRINT) std::cout << "QU: " << m << " " << k << " " << qu->GetLevel() << " " << qu->GetScalingFactor() << " " << qu->GetNoiseScaleDeg() << std::endl;
         }
     }
 
@@ -709,55 +839,76 @@ Ciphertext<DCRTPoly> InnerEvalChebyshevPS(ConstCiphertext<DCRTPoly>& x, const st
         s2.back() = 1;
 
         if (Degree(s2) > k) {
-            su = InnerEvalChebyshevPS(x, s2, k, m - 1, T, T2);
+            su = InnerEvalChebyshevPS(x, s2, k, m - 1, T, T2, level_offset + 1);
         }
         else {
-            // the highest order coefficient will always be 1 because s2 is monic.
-            su = T[k - 1]->Clone();
-
-            // ds = k from construction
-            // perform scalar multiplication for all other terms and sum them up if there are non-zero coefficients
-            s2.resize(k);
-            if (uint32_t n = Degree(s2); n > 0)
-                cc->EvalAddInPlace(su, EvalPartialLinearWSum(T, s2, n));
-
-            // adds the free term (at x^0)
-            cc->EvalAddInPlace(su, s2.front() / 2.0);
-
-            // The number of levels of su is the same as the number of levels of T[k-1] or T[k-1] + 1. Need to reduce it to T2[m-1] + 1.
-            cc->LevelReduceInPlace(su, nullptr);
+            if (BASELINE) {
+
+                if constexpr (CHEBY_PRINT) std::cout << "s2: ";
+                if constexpr (CHEBY_PRINT) for (auto i : s2) std::cout << i << ", ";
+                if constexpr (CHEBY_PRINT) std::cout << std::endl;
+                // the highest order coefficient will always be 1 because s2 is monic.
+                su = T[k - 1]->Clone();
+                if (!BASELINE) cc->EvalMultInPlace(su, 1.0, (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset + 1);
+                // ds = k from construction
+                // perform scalar multiplication for all other terms and sum them up if there are non-zero coefficients
+                s2.resize(k);
+                if (uint32_t n = Degree(s2); n > 0)
+                    cc->EvalAddInPlace(su, EvalPartialLinearWSum(T, s2, n, BASELINE ? -1 : (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset + 1, false));
+
+                //std::cout << "SU: " << m << " " << k << " " << su->GetLevel() << " " << su->GetScalingFactor() << " " << su->GetNoiseScaleDeg() << std::endl;
+                // adds the free term (at x^0)
+                cc->EvalAddInPlace(su, s2.front() / 2.0);
+
+                // The number of levels of su is the same as the number of levels of T[k-1] or T[k-1] + 1. Need to reduce it to T2[m-1] + 1.
+                if (BASELINE) cc->LevelReduceInPlace(su, nullptr);
+            } else {
+                su = EvalPartialLinearWSumWithBias(T, s2, s2.front() / (VectorDataType)2.0, (uint32_t) s2.size()-1,   (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset + 1, false);
+            }
+            if constexpr (CHEBY_PRINT) std::cout << "SU: " << m << " " << k << " " << su->GetLevel() << " " << su->GetScalingFactor() << " " << su->GetNoiseScaleDeg() << std::endl;
         }
     }
 
     if (uint32_t n = Degree(divcs->q); n >= 1) {
-        if (n == 1) {
-            if (IsNotEqualOne(divcs->q[1])) {
-                cu = cc->EvalMult(T.front(), divcs->q[1]);
-                cc->ModReduceInPlace(cu);
+        if (BASELINE) {
+            if (n == 1) {
+                if (IsNotEqualOne(divcs->q[1])) {
+                    cu = cc->EvalMult(T.front(), divcs->q[1], BASELINE ? -1 : (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset);
+                    cc->ModReduceInPlace(cu);
+                } else {
+                    cu = T.front()->Clone();
+                }
+                if constexpr (CHEBY_PRINT) std::cout << "CU simple: " << m << " " << k << " " << cu->GetLevel() << " " << cu->GetScalingFactor() << " " << cu->GetNoiseScaleDeg() << std::endl;
             }
             else {
-                cu = T.front()->Clone();
+                cu = EvalPartialLinearWSum(T, divcs->q, n, BASELINE ? -1 : (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset);
             }
-        }
-        else {
-            cu = EvalPartialLinearWSum(T, divcs->q, n);
-        }
+            if constexpr (CHEBY_PRINT) std::cout << "divcs->q: ";
+            if constexpr (CHEBY_PRINT) for (auto i : divcs->q) std::cout << i << ", ";
+            if constexpr (CHEBY_PRINT) std::cout << std::endl;
+
+            // adds the free term (at x^0)
+            cc->EvalAddInPlace(cu, divcs->q.front() / 2.0);
 
-        // adds the free term (at x^0)
-        cc->EvalAddInPlace(cu, divcs->q.front() / 2.0);
+            // Need to reduce levels up to the level of T2[m-1].
+            if (BASELINE) {
+                uint32_t cd =
+                    std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
+                cc->LevelReduceInPlace(cu, nullptr, (T2[m - 1]->GetLevel() - cu->GetLevel()) / cd);
+            }
+        } else {
+            cu = EvalPartialLinearWSumWithBias(T, divcs->q, divcs->q.front() / (VectorDataType)2.0,  divcs->q.size()-1, (int)T2[m-1]->GetLevel() - (T2[m-1]->GetNoiseScaleDeg() == 1) + level_offset);
+        }
 
-        // Need to reduce levels up to the level of T2[m-1].
-        uint32_t cd =
-            std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
-        cc->LevelReduceInPlace(cu, nullptr, (T2[m - 1]->GetLevel() - cu->GetLevel()) / cd);
+        if constexpr (CHEBY_PRINT) std::cout << "CU: " << m << " " << k << " " << cu->GetLevel() << " " << cu->GetScalingFactor() << " " << cu->GetNoiseScaleDeg() << std::endl;
     }
 
     cu = cu ? cc->EvalAdd(T2[m - 1], cu) : cc->EvalAdd(T2[m - 1], divcs->q.front() / 2.0);
-
-    auto result = cc->EvalMult(cu, qu);
-    cc->ModReduceInPlace(result);
-    cc->EvalAddInPlace(result, su);
-    return result;
+    cc->EvalMultMutableInPlace(cu, qu);
+    if (BASELINE) cc->ModReduceInPlace(cu);
+    cc->EvalAddInPlace(cu, su);
+    if constexpr (CHEBY_PRINT) std::cout << "M: " << m << " " << k << " " << cu->GetLevel() << " " << cu->GetScalingFactor() << " " << cu->GetNoiseScaleDeg() << std::endl;
+    return cu;
 }
 
 std::shared_ptr<seriesPowers<DCRTPoly>> internalEvalChebyPolysPS(ConstCiphertext<DCRTPoly>& x, uint32_t degree,
@@ -774,17 +925,34 @@ std::shared_ptr<seriesPowers<DCRTPoly>> internalEvalChebyPolysPS(ConstCiphertext
         // no linear transformation is needed if a = -1, b = 1
         // T_1(y) = y
         T[0] = x->Clone();
-    }
-    else {
+    } else {
         // linear transformation is needed
         double alpha = 2 / (b - a);
         double beta  = a * alpha;
 
-        T[0] = cc->EvalMult(x, alpha);
-        cc->ModReduceInPlace(T[0]);
-        cc->EvalAddInPlace(T[0], -1.0 - beta);
+        if (BASELINE) {
+            T[0] = cc->EvalMult(x, alpha);
+            cc->ModReduceInPlace(T[0]);
+            cc->EvalAddInPlace(T[0], -1.0 - beta);
+        } else {
+            if (!IsNotEqualNegOne(alpha)) {
+                T[0] = cc->EvalNegate(x);
+            } else if (IsNotEqualOne(alpha)) {
+                T[0] = cc->EvalMult(x, alpha);
+            }
+            if (IsNotEqualZero(-1.0-beta))
+                cc->EvalAddInPlace(T[0], -1.0 - beta);
+
+            if(T[0]->GetNoiseScaleDeg() == 2) {
+                uint32_t compositeDegree =
+            std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
+                cc->GetScheme()->ModReduceInternalInPlace(T[0], compositeDegree);
+            }
+        }
+
     }
 
+    std::vector<Ciphertext<DCRTPoly>> T2(m);
     // Computes Chebyshev polynomials up to degree k
     // for y: T_1(y) = y, T_2(y), ... , T_k(y)
     // uses binary tree multiplication
@@ -793,30 +961,46 @@ std::shared_ptr<seriesPowers<DCRTPoly>> internalEvalChebyPolysPS(ConstCiphertext
             // compute T_{2i+1}(y) = 2*T_i(y)*T_{i+1}(y) - y
             T[i - 1] = cc->EvalMult(T[i / 2 - 1], T[i / 2]);
             cc->EvalAddInPlaceNoCheck(T[i - 1], T[i - 1]);
-            cc->ModReduceInPlace(T[i - 1]);
+            if (BASELINE) {
+                cc->ModReduceInPlace(T[i - 1]);
+            }
+            if (!BASELINE) {
+                uint32_t compositeDegree =
+        std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
+                cc->GetScheme()->ModReduceInternalInPlace(T[i-1], compositeDegree);
+            }
             cc->EvalSubInPlace(T[i - 1], T[0]);
         }
         else {
             // compute T_{2i}(y) = 2*T_i(y)^2 - 1
             T[i - 1] = cc->EvalSquare(T[i / 2 - 1]);
+
             cc->EvalAddInPlaceNoCheck(T[i - 1], T[i - 1]);
-            cc->ModReduceInPlace(T[i - 1]);
-            cc->EvalAddInPlace(T[i - 1], -1.0);
+            if (BASELINE) {
+                cc->ModReduceInPlace(T[i - 1]);
+                cc->EvalAddInPlace(T[i - 1], -1.0);
+            } else {
+                cc->EvalAddInPlace(T[i - 1], -1.0);
+                uint32_t compositeDegree =
+        std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(x->GetCryptoParameters())->GetCompositeDegree();
+                cc->GetScheme()->ModReduceInternalInPlace(T[i-1], compositeDegree);
+            }
         }
     }
 
-    const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(T[k - 1]->GetCryptoParameters());
-    if (cryptoParams->GetScalingTechnique() == FIXEDMANUAL) {
-        // brings all powers of x to the same level
-        for (uint32_t i = 1; i < k; ++i)
-            cc->LevelReduceInPlace(T[i - 1], nullptr, T[k - 1]->GetLevel() - T[i - 1]->GetLevel());
-    }
-    else {
-        for (uint32_t i = 1; i < k; ++i)
-            cc->GetScheme()->AdjustLevelsAndDepthInPlace(T[i - 1], T[k - 1]);
+    if (BASELINE) {
+        const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(T[k - 1]->GetCryptoParameters());
+        if (cryptoParams->GetScalingTechnique() == FIXEDMANUAL) {
+            // brings all powers of x to the same level
+            for (uint32_t i = 1; i < k; ++i)
+                cc->LevelReduceInPlace(T[i - 1], nullptr, T[k - 1]->GetLevel() - T[i - 1]->GetLevel());
+        }
+        else {
+            for (uint32_t i = 1; i < k; ++i)
+                cc->GetScheme()->AdjustLevelsAndDepthInPlace(T[i - 1], T[k - 1]);
+        }
     }
 
-    std::vector<Ciphertext<DCRTPoly>> T2(m);
     // T2[0] is used as a placeholder
     T2[0] = T.back();
 
@@ -827,9 +1011,9 @@ std::shared_ptr<seriesPowers<DCRTPoly>> internalEvalChebyPolysPS(ConstCiphertext
         // Compute the Chebyshev polynomials T_k(y), T_{2k}(y), T_{4k}(y), ... , T_{2^{m-1}k}(y)
         T2[i] = cc->EvalSquare(T2[i - 1]);
         cc->EvalAddInPlaceNoCheck(T2[i], T2[i]);
-        cc->ModReduceInPlace(T2[i]);
+        if (BASELINE) cc->ModReduceInPlace(T2[i]);
         cc->EvalAddInPlace(T2[i], -1.0);
-
+        if (!BASELINE) cc->ModReduceInPlace(T2[i]);
         // compute T_{k(2*m - 1)} = 2*T_{k(2^{m-1}-1)}(y)*T_{k*2^{m-1}}(y) - T_k(y)
         T2km1 = cc->EvalMult(T2km1, T2[i]);
         cc->EvalAddInPlaceNoCheck(T2km1, T2km1);
@@ -858,7 +1042,10 @@ Ciphertext<DCRTPoly> internalEvalChebyshevSeriesPSWithPrecomp(const std::shared_
     f2.resize(2 * k2m2k + k + 1);
     f2.back() = 1;
 
-    return T[0]->GetCryptoContext()->EvalSub(InnerEvalChebyshevPS(T[0], f2, k, m, T, T2), T2km1);
+    auto aux = InnerEvalChebyshevPS(T[0], f2, k, m, T, T2);
+    if (!BASELINE) T[0]->GetCryptoContext()->ModReduceInPlace(aux);
+    T[0]->GetCryptoContext()->EvalSubInPlace(aux, T2km1);
+    return aux;
 }
 
 std::shared_ptr<seriesPowers<DCRTPoly>> AdvancedSHECKKSRNS::EvalChebyPolys(ConstCiphertext<DCRTPoly>& x,
